### ISSUE TITLE
Autobatch segmentation fault 

### ISSUE THREAD

**Issue by afonsoamendes** [clab/dynet#1559]
Hi,

I have a problem using autobatching (it does not happen without).

Here is gdb output 
`
Thread 1 "HNMTrainer" received signal SIGSEGV, Segmentation fault.
0x00007ffff51bbb9a in dynet::BatchedExecutionEngine::backward (this=<optimized out>, from_where=45896, full=<optimized out>)
     at /home/amm/pba44c/Users/amm/Source/Repos/HierarchicalNestedNER/dynet/dynet/exec.cc:1074
1074                Tensor my_ndEdf = *xs[ai];

(gdb) p ai
$4 = 3
(gdb) p xs
$5 = std::vector of length 3, capacity 118 = {0x7ffe9e879090, 0x7ffe9e8790d0, 0x7fff14aed140}

`

The other variables were optimized out, this happens after 51000 updates and if I change the batch size it can happen much latter.

Now I can put the execute the code to this point jumping over the previous updates anf get to this state quickly.

If needed I can compile dynet without optimizations and do some guided debugging.

Thanks

Afonso

**Comment by neubig** [comment]
Is this in C++? I think the first thing that would be helpful is if you could run cg.print_graphviz() to output the shape of the graph. It would be especially helpful if it's not a big graph.

**Comment by afonsoamendes** [comment]
Yes it is in C++. I think that the graph is quite big, but here it is.
[graph.txt](https://github.com/clab/dynet/files/3262646/graph.txt)
